### PR TITLE
PV61 hash downloading fixes

### PR DESCRIPTION
**Pull request by arkpar**

### CODE REVIEW

File: libethereum/EthereumHost.cpp [hunk 1]
```diff
@@ -91,7 +91,10 @@ void EthereumHost::doWork()
 	bool netChange = ensureInitialised();
 	auto h = m_chain.currentHash();
 	// If we've finished our initial sync (including getting all the blocks into the chain so as to reduce invalid transactions), start trading transactions & blocks
-	if (!isSyncing() && m_chain.isKnown(m_latestBlockSent))
+	bool syncing = false;
```
Comment (gavofyork) [inline libethereum/EthereumHost.cpp]: isSyncing() is guarded by x_sync anyway. why twice?

File: libethereum/EthereumHost.cpp [hunk 2]
```diff
@@ -470,8 +444,7 @@ void EthereumHost::onPeerBlocks(EthereumPeer* _peer, RLP const& _r)
 
 void EthereumHost::onPeerNewHashes(EthereumPeer* _peer, h256s const& _hashes)
 {
-	Guard l(x_sync);
-	if (_peer->m_asking != Asking::Nothing)
+	if (isSyncing())
```
Comment (gavofyork) [inline libethereum/EthereumHost.cpp]: no guard of `x_sync` here?

File: libethereum/EthereumHost.cpp [hunk 3]
```diff
@@ -91,7 +91,10 @@ void EthereumHost::doWork()
 	bool netChange = ensureInitialised();
 	auto h = m_chain.currentHash();
 	// If we've finished our initial sync (including getting all the blocks into the chain so as to reduce invalid transactions), start trading transactions & blocks
-	if (!isSyncing() && m_chain.isKnown(m_latestBlockSent))
+	bool syncing = false;
+	DEV_GUARDED(x_sync)
```
Comment (gavofyork) [inline libethereum/EthereumHost.cpp]: why move x_sync guard out of isSyncing?

**Comment by gavofyork** [comment]
need to explain this patch a little more - i don't see the need to remove x_sync from isSyncing.

File: libethereum/EthereumHost.h [hunk 4]
```diff
@@ -94,6 +92,7 @@ class EthereumHost: public p2p::HostCapability<EthereumPeer>, Worker
 	std::pair<std::vector<std::shared_ptr<EthereumPeer>>, std::vector<std::shared_ptr<EthereumPeer>>> randomSelection(unsigned _percent = 25, std::function<bool(EthereumPeer*)> const& _allow = [](EthereumPeer const*){ return true; });
 	void forEachPeerPtr(std::function<void(std::shared_ptr<EthereumPeer>)> const& _f) const;
 	void forEachPeer(std::function<void(EthereumPeer*)> const& _f) const;
+	bool isSyncingInternal() const;
```
Comment (gavofyork) [inline libethereum/EthereumHost.h]: best renamed to `isSyncing_UNSAFE` as is used in other classes (TransactionQueue/BlockQueue, iirc)

**Comment by arkpar** [comment]
Added some comments to isSyncing_UNSAFE

File: libethereum/EthereumPeer.cpp [hunk 5]
```diff
@@ -88,8 +86,7 @@ void EthereumPeer::setIdle()
 
 void EthereumPeer::requestStatus()
 {
-	if (m_asking != Asking::Nothing)
-	clog(NetWarn) << "Bad state: requesting state should be the first action";
+	assert(m_asking == Asking::Nothing);
```
Comment (gavofyork) [inline libethereum/EthereumPeer.cpp]: this is /very/ harsh; why bother crashing the app in the case it's wrong rather than attempting to recover?

Comment (arkpar) [inline libethereum/EthereumPeer.cpp]: This is programming invariant, not a runtime error we can recover from. requestStatus is private and called from constructor only. If assert fails here this is bad programming, and better Fail Fast

Comment (gavofyork) [inline libethereum/EthereumPeer.cpp]: OK, yes, that's true here.

File: libethereum/EthereumPeer.cpp [hunk 6]
```diff
@@ -106,22 +103,22 @@ void EthereumPeer::requestStatus()
 
 void EthereumPeer::requestHashes()
 {
-	if (m_asking == Asking::Blocks)
-		return;
+	assert(m_asking == Asking::Nothing);
```
Comment (gavofyork) [inline libethereum/EthereumPeer.cpp]: this is /very/ harsh; why bother crashing the app in the case, rather than attempting to recover?

**Comment by gavofyork** [comment]
will merge anyway, though i'd prefer all *dynamic* state assumptions/requirements to be recoverable.

**Pull request merged**